File: misc/experiments/strucs/TMBlist.R
library(TMB)

set.seed(101)
data <- list(
  MatrixList = list(matrix(rnorm(9), 3, 3)),
  VectorList = list(rnorm(3))
)
param = list(theta = rnorm(1))

compile("TMBlist.cpp")
dd <- dyn.load(dynlib("TMBlist"))
dlls <- getLoadedDLLs()
getDLLRegisteredRoutines(dd)
## hmmm ...

## debug(MakeADFun)
obj <- MakeADFun(
  data = data,
  parameters = param,
  DLL = "TMBlist"
)

obj$fn() == (data$MatrixList[[1]] * data$VectorList[[1]] * param$theta)[1]
obj$gr() == (data$MatrixList[[1]] * data$VectorList[[1]])[1]

## current status
## crashes R unpredictably
##  sometimes? prints debugging statements, sometimes doesn't
## crashes on "retape(set.defaults = TRUE)" (i.e. first call to
##   DLL ...)

File: misc/experiments/strucs/TMBlist.cpp
#include <TMB.hpp>
#include <iostream>
#include <string>

template<class Type>
Type objective_function<Type>::operator() ()
{

  std::cout << "a\n";

  SEXP MatrixList(getListElement(TMB_OBJECTIVE_PTR -> data, "MatrixList"));
  SEXP VectorList(getListElement(TMB_OBJECTIVE_PTR -> data, "VectorList"));
  PARAMETER(theta);

  std::cout << "b\n";
  // for (int i = 0; i<(int)Rf_length(MatrixList); i++) {
    matrix<Type> m(asMatrix<Type>(VECTOR_ELT(MatrixList, 0)));
    vector<Type> v(asVector<Type>(VECTOR_ELT(VectorList, 0)));

    std::cout << "c\n";
    //vector<Type> v2 = m * v * theta;
    Type yy = m.coeff(0, 0) * v[0] * theta;

    std::cout << "d\n";
    //}
  //Type yy = 0;
  return yy;
}
